core/src/main/scala/org/apache/spark/shuffle/hash/HashShuffleManager.scala

```diff
@@ -50,15 +53,30 @@ private[spark] class HashShuffleManager(conf: SparkConf) extends ShuffleManager
   }
 
   /** Get a writer for a given partition. Called on executors by map tasks. */
-  override def getWriter[K, V](handle: ShuffleHandle, mapId: Int, context: TaskContext)
-      : ShuffleWriter[K, V] = {
-    new HashShuffleWriter(
-      shuffleBlockResolver, handle.asInstanceOf[BaseShuffleHandle[K, V, _]], mapId, context)
+  override def getWriter[K, V](
+      handle: ShuffleHandle,
+      mapId: Int,
+      stageAttemptId: Int,
+      context: TaskContext): ShuffleWriter[K, V] = {
+    addShuffleAttempt(handle.shuffleId, stageAttemptId)
+    new HashShuffleWriter(shuffleBlockResolver, handle.asInstanceOf[BaseShuffleHandle[K, V, _]],
```

Comment:
Some sort of recognition at ShuffleManger.getWriter that addShuffleAttempt must occur is needed.  The bare minimum would be the additional commentary that I requested earlier; but even better would be if we can make the compiler enforce the requirement.